src/VisualStudio/Core/Def/Implementation/Experimentation/KeybindingResetDetector.cs

```diff
@@ -0,0 +1,358 @@
+        private static readonly Guid ReSharperCommandGroup = new Guid("{47F03277-5055-4922-899C-0F7F30D26BF1}");
+
+        private readonly VisualStudioWorkspace _workspace;
+        private readonly SVsServiceProvider _serviceProvider;
+
+        private IExperimentationService _experimentationService;
+        private IVsUIShell _uiShell;
+        private IOleCommandTarget _oleCommandTarget;
+        private OleComponent _oleComponent;
+
+        private bool _disposedValue = false;
+        private uint _priorityCommandTargetCookie = VSConstants.VSCOOKIE_NIL;
+
+        private bool _infoBarOpen = false;
+
+        [ImportingConstructor]
+        public KeybindingResetDetector(VisualStudioWorkspace workspace, SVsServiceProvider serviceProvider)
+        {
+            _workspace = workspace;
+            _serviceProvider = serviceProvider;
+        }
+
+        public Task Initialize()
+        {
+            // Immediately bail if the user has asked to never see this bar again.
+            if (_workspace.Options.GetOption(KeybindingResetOptions.NeverShowAgain))
+            {
+                return Task.CompletedTask;
+            }
+
+            return InvokeBelowInputPriority(InitializeCore);
+        }
+
+        private void InitializeCore()
+        {
+            AssertIsForeground();
+
+            // Ensure one of the flights is enabled, otherwise bail
+            _experimentationService = _workspace.Services.GetRequiredService<IExperimentationService>();
+            if (!_experimentationService.IsExperimentEnabled(ExternalFlightName) && !_experimentationService.IsExperimentEnabled(InternalFlightName))
+            {
+                return;
+            }
+
+            var vsShell = _serviceProvider.GetService(typeof(SVsShell)) as IVsShell;
+            var hr = vsShell.IsPackageInstalled(ReSharperPackageGuid, out int extensionEnabled);
+            if (ErrorHandler.Failed(hr))
+            {
+                FatalError.ReportWithoutCrash(Marshal.GetExceptionForHR(hr));
+                return;
+            }
+
+            var currentStatus = _workspace.Options.GetOption(KeybindingResetOptions.ReSharperStatus);
+
+            if (extensionEnabled == 0)
+            {
+                // If 0, the extension is either disabled in the extension manager, or not installed at all.
+                // If this is a change, update the status.
+                bool needsReset = _workspace.Options.GetOption(KeybindingResetOptions.NeedsReset);
+                if (currentStatus != ReSharperStatus.NotInstalledOrDisabled)
+                {
+                    // If we're going directly from Enabled->NotInstalled, we need to reset keybindings. Otherwise, the previous value of NeedsReset
+                    // is correct.
+                    var changedOptions = _workspace.Options;
+                    if (currentStatus == ReSharperStatus.Enabled)
```

Comment:
Isn't there a lot of duplicate logic here with the other function below? Can't we just make this "fetch previous status, fetch current status", and then if previous status == enabled && current status != enabled, needs reset = true? It seems like we'll get rid of lots of the logic here.